File: 01-Languages/Kotlin/07-Collections/Exercises/Exercise.kt
// Exercise: given a list of Product data classes, use filter/map/sumOf to compute the
// total price of in-stock products, and a re-usable list of just the names of in-stock products.

data class Product(val name: String, val price: Double, val inStock: Boolean)

fun totalInStockPrice(products: List<Product>): Double {
    // TODO: filter to inStock, map/sumOf price
    return 0.0
}

fun inStockNames(products: List<Product>): List<String> {
    // TODO: filter to inStock, map to name
    return emptyList()
}

fun main() {
    val products = listOf(
        Product("Widget", 9.99, true),
        Product("Gadget", 19.99, false),
        Product("Gizmo", 14.99, true),
    )
    println("total: ${totalInStockPrice(products)}") // expected: 24.98
    println("names: ${inStockNames(products)}")        // expected: [Widget, Gizmo]
}

File: 01-Languages/PHP/07-Collections/Exercises/exercise.php
<?php
declare(strict_types=1);
// Exercise: given a list of associative arrays representing products, use array_filter,
// array_map, and array_sum to compute the total price of all in-stock products, and
// return a re-indexed (array_values) list of just the names of in-stock products.

$products = [
    ["name" => "Widget", "price" => 9.99, "inStock" => true],
    ["name" => "Gadget", "price" => 19.99, "inStock" => false],
    ["name" => "Gizmo", "price" => 14.99, "inStock" => true],
];

function totalInStockPrice(array $products): float {
    // TODO: filter to inStock === true, map to "price", sum the result
    return 0.0;
}

function inStockNames(array $products): array {
    // TODO: filter to inStock === true, map to "name", re-index with array_values
    return [];
}

echo "total: " . totalInStockPrice($products) . "\n"; // expected: 24.98
echo "names: " . implode(", ", inStockNames($products)) . "\n"; // expected: Widget, Gizmo
